### PR TITLE
Expose Cache engines

### LINKED ISSUE #9741: Cache does not expose (configured) engines
* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 2.9.2 (but also applies to other versions)

Is there a reason that `Cache` does not expose the instances it holds in `static::$_engines`? For our use-case, but probably also for others, it would be very helpful to have something like this:

```php
public static function getEngine($config = 'default') {
	$settings = static::settings($config);

	if (empty($settings)) {
		return false;
	}
	if (!static::isInitialized($config)) {
		return false;
	}
	return static::$_engines[$config];
}
```

This way I can call Engine specific methods, for instance `ping` on the `Redis` object:

```php
$pong = Cache::getEngine()->ping();
```

Or `add` (with four arguments) on the `Memcache` object:

```php
$success = Cache::getEngine(Configure::read('Session.handler.config'))->add($this->__lockKey, '1', null, $this->__lockMaxWait + 1);
```

I would love to implement this feature, but before I do so I would like to discuss the idea here.

Thanks in advance.

### PR DESCRIPTION
**Pull request by tersmitten**
By implementing the `engine` method just like in version 3

Fixes #9741

### CODE REVIEW

File: lib/Cake/Test/Case/Cache/CacheTest.php [hunk 1]
```diff
@@ -538,4 +538,64 @@ public function testAdd() {
...
+ * @return void
+ */
+	public function testEngineSuccessMemcached() {
+		$this->skipIf(!class_exists('Memcached'), 'Memcached is not installed or configured properly.');
+
+		// @codingStandardsIgnoreStart
+		$socket = @fsockopen('127.0.0.1', 11211, $errno, $errstr, 1);
+		// @codingStandardsIgnoreEnd
+		$this->skipIf(!$socket, 'Memcached is not running.');
+		fclose($socket);
```
Comment (markstory) [inline lib/Cake/Test/Case/Cache/CacheTest.php]: Any reason you couldn't use `File` here? That engine always exists.

Comment (tersmitten) [inline lib/Cake/Test/Case/Cache/CacheTest.php]: No, but `File` is already tested in `testEngineSuccess` and I was interested in the `add` call in `MemcachedEngine`. Do you want me to change it?

File: lib/Cake/Test/Case/Cache/CacheTest.php [hunk 2]
```diff
@@ -538,4 +538,64 @@ public function testAdd() {
...
+
+		Cache::config('memcached', array(
+			'engine' => 'Memcached',
+			'prefix' => 'cake_',
+			'duration' => 3600
+		));
+
+		$actual = Cache::engine('memcached');
```
Comment (tersmitten) [inline lib/Cake/Test/Case/Cache/CacheTest.php]: Maybe also add `$this->assertIsA($actual, 'MemcachedEngine');`

File: lib/Cake/Test/Case/Cache/CacheTest.php (hunk 1 above)
Comment (markstory) [inline lib/Cake/Test/Case/Cache/CacheTest.php]: I only asked because it adds another dependency for this test which makes it run in fewer environments.

**Comment by tersmitten** [comment]
I think this is ready for merging. Do you agree?

**Comment by tersmitten** [comment]
Any news on this? It would be nice if this get's included in the next release.

File: lib/Cake/Cache/Cache.php [hunk 3]
```diff
@@ -616,4 +616,19 @@ public static function add($key, $value, $config = 'default') {
...
+ *
+ * @param string $config Optional string configuration name to get an engine for. Defaults to 'default'.
+ * @return bool|CacheEngine False if the engine has not been initialized else the engine
+ */
+	public static function engine($config = 'default') {
+		if (self::isInitialized($config)) {
+			return self::$_engines[$config];
+		}
+
+		return false;
```
Comment (dereuromark) [inline lib/Cake/Cache/Cache.php]: Why not just Null?

**Comment by dereuromark** [comment]
@markstory The return false should never be false IMO, always either null or the object.
It otherwise creates the same issues when assigning it to class properties as we currently face in 3.x.

**Comment by markstory** [comment]
Ok. I'll change it then.

**Comment by markstory** [comment]
Updated in bbb87b3e87d751fef6c2d202fc38a975a5c95673

**Comment by tersmitten** [comment]
Thanks @markstory, @dereuromark

File: lib/Cake/Cache/Cache.php (hunk 3 above)
Comment (tersmitten) [inline lib/Cake/Cache/Cache.php]: `null` is also fine by me, I think I copied it from another method in this class.

**Pull request merged**